src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramUniformFactory.cpp

```diff
@@ -216,6 +216,46 @@ class UScreenSpaceTriangleInfo : public UniformGroup
 	iUniform uScreenSpaceTriangle;
 };
 
+class URasterInfo : public UniformGroup {
+public:
+	URasterInfo(GLuint _program) {
+		LocateUniform(uVertexOffset);
+		LocateUniform(uTexCoordOffset);
+	}
+
+	void update(bool _force) override {
+		const bool isNativeRes = config.frameBufferEmulation.nativeResFactor == 1 && config.video.multisampling == 0;
+		const bool isTexRect = dwnd().getDrawer().getDrawingState() == DrawingState::TexRect;
+		float scale[2] = { 0.0f, 0.0f };
+		if (config.frameBufferEmulation.nativeResFactor != 0) {
+			scale[0] = scale[1] = static_cast<float>(config.frameBufferEmulation.nativeResFactor);
+		}
+		else {
+			scale[0] = dwnd().getScaleX();
+			scale[1] = dwnd().getScaleY();
+		}
+		const float vertexOffset = isNativeRes ? 0.5f : 0.0f;
+		float texCoordOffset[2] = { 0.0f, 0.0f };
+		if (isTexRect && !isNativeRes) {
+			if (gDP.otherMode.textureFilter != G_TF_POINT && gDP.otherMode.cycleType != G_CYC_COPY) {
```

Comment:
Please add comments, why you use texCoordOffset in non-native res mode and why is that difference in code between gDP.otherMode.textureFilter modes. You may put a reference to your explanations in the #2324 GitHub task.

Reply:
Done. Please check if it satifactory. I can shorten it if you feel it is too much.